src/accountsdb/download.zig

```diff
@@ -0,0 +1,652 @@
+
+        var max_inc_hash: ?SlotAndHash = null;
+        for (snapshot_hashes.incremental) |inc_hash| {
+            if (max_inc_hash == null or inc_hash.slot > max_inc_hash.?.slot) {
+                max_inc_hash = inc_hash;
+            }
+        }
+
+        // dont try to download from a slow peer
+        for (blacklist) |black_list_peers| {
+            if (black_list_peers.equals(&peer_contact_info.pubkey)) {
+                result.is_blacklist += 1;
+                continue :search_loop;
+            }
+        }
+
+        valid_peers.appendAssumeCapacity(.{
+            // NOTE: maybe we need to deep clone here due to arraylist sockets?
+            .contact_info = peer_contact_info.*,
+            .full_snapshot = snapshot_hashes.full,
+            .inc_snapshot = max_inc_hash,
+        });
+    }
+    result.is_valid = valid_peers.items.len;
+
+    return result;
+}
+
+/// uses swap remove to remove peers that do not have a hash which is trusted
+pub fn removeUnTrustedSnapshotPeers(
+    allocator: std.mem.Allocator,
+    table: *const GossipTable,
+    trusted_validators: std.ArrayList(Pubkey),
+    available_snapshot_peers: *std.ArrayList(PeerSnapshotHash),
+) !void {
+    var trusted_snapshot_hashes = std.AutoHashMap(
+        SlotAndHash, // full snapshot hash
+        std.AutoHashMap(SlotAndHash, void), // set of incremental snapshots
+    ).init(allocator);
+    defer trusted_snapshot_hashes.deinit();
+
+    // get the hashes of trusted validators
+    var trusted_count: usize = 0;
+    for (trusted_validators.items) |trusted_validator| {
+        const gossip_data = table.get(.{ .SnapshotHashes = trusted_validator }) orelse continue;
+        const trusted_hashes = gossip_data.value.data.SnapshotHashes;
+        trusted_count += 1;
+
+        var max_inc_hash: ?SlotAndHash = null;
+        for (trusted_hashes.incremental) |inc_hash| {
+            if (max_inc_hash == null or inc_hash.slot > max_inc_hash.?.slot) {
+                max_inc_hash = inc_hash;
+            }
+        }
+
+        // track the full and incremental
+        var r = try trusted_snapshot_hashes.getOrPut(trusted_hashes.full);
+        if (max_inc_hash) |inc_snapshot| {
+            if (!r.found_existing) {
+                r.value_ptr.* = std.AutoHashMap(SlotAndHash, void).init(allocator);
+            }
+            try r.value_ptr.put(inc_snapshot, {});
+        }
+    }
+
+    if (trusted_count == 0) {
+        available_snapshot_peers.clearRetainingCapacity();
+        return;
+    }
+
+    // remove peers that do not have a trusted hash
+    var peer_index: usize = 0;
+    for (0..available_snapshot_peers.items.len) |_| {
+        const peer = available_snapshot_peers.items[peer_index];
+        var should_remove = false;
+        // full snapshot must be trusted
+        if (trusted_snapshot_hashes.getEntry(peer.full_snapshot)) |entry| {
+            // if we have an incremental snapshot
+            if (peer.inc_snapshot) |inc_snapshot| {
```

Comment:
This trusts the peer to decide whether it needs to bring along an incremental snapshot. It only restricts the peer's incremental snapshot if it offers one. What if all the trusted validators specify incremental snapshots, but this peer specifies none?

Reply:
if this peer specifies none then we wont attempt to download it from them so its ok -- see `// no incremental snapshot, thats ok` comment